src/vcpkg/base/checks.cpp

```diff
@@ -13,17 +13,22 @@ namespace
     {
         return LocalizedString::from_raw(fmt::format("{}: ", line_info));
     }
-    void (*g_shutdown_handler)() = nullptr;
+
+    std::atomic<void (*)(void*)> g_shutdown_handler;
+    void* g_shutdown_parameter; // guarded by g_shutdown_handler
 }
 
 namespace vcpkg
 {
-    void Checks::register_global_shutdown_handler(void (*func)())
+    void Checks::register_global_shutdown_handler(void (*func)(void*), void* shutdown_parameter)
     {
-        if (g_shutdown_handler)
+        if (g_shutdown_handler.exchange(func))
+        {
             // Setting the handler twice is a program error. Terminate.
             std::abort();
-        g_shutdown_handler = func;
+        }
+
+        g_shutdown_parameter = shutdown_parameter;
```

Comment:
It must be set after the exchange because the exchange is what tells this thread that it owns `g_shutdown_parameter`.

I guess I can see if I can eliminate this thing since the only caller is main so globals there aren't *too* bad.